[APM] Add obfuscation for OpenSearch traces

ichinaski opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

This PR adds obfuscation support for OpenSearch statements within span metadata.

OpenSearch spans have a similar structure to ElasticSearch ones, and the main difference lies in the `opensearch.*` vs `elasticsearch.*` tags. For example, `opensearch.body` vs `elasticsearch.body`.

While the tag keys differ, the span type can be either `elasticsearch` (dd-trace-js) or `opensearch` (dd-trace-java). 

The configuration is done in the same way as elasticsearch one, via environment variables (see releasenotes).

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

Jira ticket: https://datadoghq.atlassian.net/browse/APMSP-914
Support ticket: https://datadoghq.atlassian.net/browse/FRAPMS-3979

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

ichinaski committed:
[APM] Add obfuscation for OpenSearch traces

This PR adds obfuscation support for OpenSearch.

OpenSearch spans have a similar structure to ElasticSearch ones, and the
main difference lies in the `opensearch.*` vs `elasticsearch.*` tags.
For example, `opensearch.body` vs `elasticsearch.body`.

While the tag keys differ, the span type is the same for both:
`elasticsearch`.

ajgajg1134 commented on pkg/trace/config/config.go:
```diff
@@ -85,6 +85,9 @@ type ObfuscationConfig struct {
 	ES obfuscate.JSONConfig `mapstructure:"elasticsearch"`
 
+	// OpenSearch holds the obfuscation configuration for OpenSearch bodies.
+	OpenSearch obfuscate.JSONConfig `mapstructure:"opensearch"`
```
I think we also need some changes in `comp/trace/config/setup.go` to setup the configuration based on the incoming yaml (see the example on line 411 from elasticsearch)

ichinaski replied:
Thanks for the pointer. I updated the setup based on the config.

ichinaski committed:
[APM] Setup opensearch obfuscation from config

ichinaski committed:
[APM] Fix OpenSearch configuration setup

Some elasticsearch values were copy-pasted. Fixed.

ajgajg1134 reviewed (commented):
Overall looks good to me! I think it just needs a release-note so folks know that we're adding this and it's enabled by default

ichinaski committed:
[APM] Include opensearch type on obfuscation filter

Opensearch integrations have an inconsisten span type:

- JS integration uses `elasticsearch` type
- Java integration uses `opensearch` type

The obfuscation logic will therefore look for either of those, and the
apply the obfuscation based on the existing configuration and the
presence of either `elasticsearch.body` or `opensearch.body`

ichinaski committed:
[APM] Add opensearch configuration env vars

ichinaski committed:
[APM] Add releasenotes for opensearch obfuscation

misteriaud reviewed (approved):
LGTM for code owned by ASC !

knusbaum reviewed (commented):
Do we need to add this to the config template like the elasticsearch?
https://github.com/DataDog/datadog-agent/blob/30fa0a20418e8f6e3c1cacb9ca78441442361709/pkg/config/config_template.yaml#L1127-L1138

rtrieu commented on releasenotes/notes/add-opensearch-obfuscation-a26c678785bce376.yaml:
```diff
@@ -0,0 +1,19 @@
+---
+enhancements:
+  - |
+    APM: Add obfuscation support for opensearch statements within span metadata. This feature works in the same way as the existing elasticsearch one, and it's enabled by default. It is configured by binding ``apm_config.obfuscation.opensearch.*`` parameters to new obfuscation environment variables. In particular, bind:
```
```suggestion
    APM: Add obfuscation support for OpenSearch statements within span metadata. This feature works in the same way as the existing Elasticsearch one, and is enabled by default. It is configured by binding ``apm_config.obfuscation.opensearch.*`` parameters to new obfuscation environment variables. In particular, bind:
```

rtrieu reviewed (changes requested):
Minor feedback but otherwise looks good.

ichinaski committed:
[APM] Update releasenotes/notes/add-opensearch-obfuscation-a26c678785bce376.yaml

Co-authored-by: Rosa Trieu <107086888+rtrieu@users.noreply.github.com>

rtrieu reviewed (approved):
LGTM

ichinaski committed:
[APM] Add OpenSearch env vars config to template

ichinaski commented:
> Do we need to add this to the config template like the elasticsearch?
> 
> https://github.com/DataDog/datadog-agent/blob/30fa0a20418e8f6e3c1cacb9ca78441442361709/pkg/config/config_template.yaml#L1127-L1138

Good point! I added this in a new commit.

knusbaum reviewed (approved):
LGTM

ichinaski commented:
/merge
